Update trackers to use manifest and handle tar archives

michellewang opened the pull request:
Closes #180, closes #174, closes #97 (except for the `--check-for-update` flag)

With a minor bug fix that made the script write new bagel files even if they didn't change from the previous version.

Some notes:
- Added more logic to handle the fMRIPrep 20.2.7 results being in `output/fmriprep` instead of just `output`. So @nikhil153 depending on your setup for QPN/NIMHANS we might need to put things back into an `fmriprep` subdirectory
- I have a personal preference for using the version name without a preceding `v`, which is how I have coded the Boutiques thing to do for now. This PR makes the tracker compatible with the current Boutiques runner outputs. So this will break things with older runscripts unless we rename the version directories (e.g. `v20.2.7` -> `20.2.7`). @nikhil153 let me know what you think about that

nikhil153 commented:
Both of these issues are I think user- vs dev- friendliness tradeoff :) 

- From user perspective, I think it's better to standardize root output dir across pipelines. Which is why I was manually moving the data in the legacy `fmriprep20.2.7` which creates confusing `fmriprep/v6.0.1/output/fmriprep` path. If manual moving is annoying (which it is) maybe we should just check and support both output paths?
 
- Why do you not like the `v` prefix? I don't have strong preference here. I was going with what I usually see in the user facing doc pages (see [example](https://fmriprep.org/en/stable/)) to make it intuitive for new users. But it shouldn't be difficult to change it.

nikhil153 commented on nipoppy/trackers/run_tracker.py:
```diff
@@ -34,6 +32,13 @@
+    "fmriprep": ["anat"],
+    "mriqc": ["anat"],
+    "tractoflow": ["anat", "dwi"],
+}
```
Should this go in global_config?

michellewang replied:
I was thinking this should go in the individual trackers instead? It would be the datatypes required for the `pipeline_complete` part of the pipeline to run

nikhil153 replied:
Individual trackers makes sense!

nikhil153 commented on nipoppy/trackers/run_tracker.py:
```diff
@@ -154,30 +156,41 @@ def run(global_configs, dash_schema_file, pipelines, session_id="ALL", run_id="1
                 elif pipeline in BIDS_PIPES:
-                    subject_dir = f"{DATASET_ROOT}/derivatives/{pipeline}/v{version}/output/{bids_id}" 
+                    subject_dir = f"{DATASET_ROOT}/derivatives/{pipeline}/{version}/output/{bids_id}" 
+                    if pipeline == "fmriprep" and version == "20.2.7":  # NOTE temporary solution while we refactor tracker configs to be version-specific
```
I think it would be more robust if we just check both output paths instead of being very specific to a given fmriprep version string.

nikhil153 commented on nipoppy/trackers/run_tracker.py:
```diff
@@ -154,30 +156,41 @@ def run(global_configs, dash_schema_file, pipelines, session_id="ALL", run_id="1
                     
                 subject_ses_dir_status = Path(subject_ses_dir).is_dir()
-                logger.debug(f"subject_dir:{subject_ses_dir_status}, dir_status: {subject_ses_dir_status}")
+                subject_ses_tar_status = Path(subject_ses_dir).with_suffix('.tar').is_file()
```
Will always be `tar` or `tar.gz` is also possible?

nikhil153 replied:
Also if it doesn't slow down things too much, might be worth adding the `tarfile.is_tarfile()` check.

michellewang replied:
Oh I didn't know about `tarfile.is_tarfile()` 😅  but I just tried it and it seems like it is slowing things down a fair bit 😞 (PPMI fMRIPrep: ~6 minutes with `is_tarfile`, ~1 minute with just checking file existence)

michellewang commented:
@nikhil153 added the additional check for `.tar.gz` file extension

> From user perspective, I think it's better to standardize root output dir across pipelines. Which is why I was manually moving the data in the legacy fmriprep20.2.7 which creates confusing fmriprep/v6.0.1/output/fmriprep path. If manual moving is annoying (which it is) maybe we should just check and support both output paths?

Ok, I changed it to check both output paths for now, but it is very messy (since it affects the path to the subject directory, the path to the session directory, and the path to the potential tarred archive). So I still think we should eventually have a more user-configurable way to get paths to track for different versions.

> Why do you not like the v prefix? I don't have strong preference here. I was going with what I usually see in the user facing doc pages (see [example](https://fmriprep.org/en/stable/)) to make it intuitive for new users. But it shouldn't be difficult to change it.

I think it might be a Pypi thing to have the `v` prefix... I just don't think it's necessary 😕 

Also added the `HAS_DATATYPE__<datatype>` columns since that's easy to do now that we have the manifest. Mentioned by @alyssadai today when talking about the dashboard.

michellewang commented:
@nikhil153 could you take another look? The Neurobagel team will try to get a someone at the Douglas to run the BIDS tracker soon, and it would be good if they only need the manifest file (not the doughnut) for that

nikhil153 commented on nipoppy/trackers/run_tracker.py:
```diff
@@ -192,23 +226,33 @@ def run(global_configs, dash_schema_file, pipelines, session_id="ALL", run_id="1
+                        _df.loc[bids_id, name] = INCOMPLETE
+
+                        # check that all required datatypes are available
+                        for datatype in PIPELINE_REQUIRED_DATATYPES[pipeline]:
```
This branching feels a bit odd - shouldn't the "availability" of participant be verified before checking for "incomplete" status?

nikhil153 commented:
>Ok, I changed it to check both output paths for now, but it is very messy (since it affects the path to the subject directory, the path to the session directory, and the path to the potential tarred archive). So I still think we should eventually have a more user-configurable way to get paths to track for different versions.

Yes made a comment about that. Other than that it's good to go!

michellewang commented:
@nikhil153 I changed the branching logic for `UNAVAILABLE` status based on the manifest's `datatypes` column like we discussed this morning
